apps/cowswap-frontend/src/modules/swap2/containers/SwapWidget/index.tsx

```diff
@@ -0,0 +1,198 @@
+    wrapCallback,
+    directSwapCallback: doTrade.callback,
+    hasEnoughWrappedBalanceForSwap,
+  })
+
+  const inputCurrencyInfo: CurrencyInfo = {
+    field: Field.INPUT,
+    currency: inputCurrency,
+    amount: inputCurrencyAmount,
+    isIndependent: isSellTrade,
+    balance: inputCurrencyBalance,
+    fiatAmount: inputCurrencyFiatAmount,
+    receiveAmountInfo: !isSellTrade ? receiveAmountInfo : null,
+  }
+
+  const outputCurrencyInfo: CurrencyInfo = {
+    field: Field.OUTPUT,
+    currency: outputCurrency,
+    amount: outputCurrencyAmount,
+    isIndependent: !isSellTrade,
+    balance: outputCurrencyBalance,
+    fiatAmount: outputCurrencyFiatAmount,
+    receiveAmountInfo: isSellTrade ? receiveAmountInfo : null,
+  }
+  const inputCurrencyPreviewInfo = {
+    amount: inputCurrencyInfo.amount,
+    fiatAmount: inputCurrencyInfo.fiatAmount,
+    balance: inputCurrencyInfo.balance,
+    label: isSellTrade ? 'Sell amount' : 'Expected sell amount',
+  }
+
+  const outputCurrencyPreviewInfo = {
+    amount: outputCurrencyInfo.amount,
+    fiatAmount: outputCurrencyInfo.fiatAmount,
+    balance: outputCurrencyInfo.balance,
+    label: isSellTrade ? 'Receive (before fees)' : 'Buy exactly',
+  }
+
+  const rateInfoParams = useRateInfoParams(inputCurrencyInfo.amount, outputCurrencyInfo.amount)
+
+  const buyingFiatAmount = useMemo(
+    () => (isSellTrade ? outputCurrencyInfo.fiatAmount : inputCurrencyInfo.fiatAmount),
+    [isSellTrade, outputCurrencyInfo.fiatAmount, inputCurrencyInfo.fiatAmount],
+  )
+
+  const slots: TradeWidgetSlots = {
+    topContent,
+    settingsWidget: (
+      <SettingsTab
+        recipientToggleState={recipientToggleState}
+        hooksEnabledState={hooksEnabledState}
+        deadlineState={deadlineState}
+      />
+    ),
+    bottomContent: useCallback(
+      (tradeWarnings: ReactNode | null) => {
+        return (
+          <>
+            {bottomContent}
+            <TradeRateDetails
+              isTradePriceUpdating={isRateLoading}
+              rateInfoParams={rateInfoParams}
+              deadline={deadlineState[0]}
+            />
+            <Warnings buyingFiatAmount={buyingFiatAmount} />
+            {tradeWarnings}
+            <TradeButtons
+              isTradeContextReady={doTrade.contextIsReady}
+              openNativeWrapModal={openNativeWrapModal}
+              hasEnoughWrappedBalanceForSwap={hasEnoughWrappedBalanceForSwap}
+            />
+          </>
+        )
+      },
+      [
+        doTrade.contextIsReady,
+        isRateLoading,
+        rateInfoParams,
+        deadlineState,
+        bottomContent,
+        buyingFiatAmount,
+        hasEnoughWrappedBalanceForSwap,
+        openNativeWrapModal,
+      ],
+    ),
+  }
+
+  const params = {
+    compactView: true,
+    enableSmartSlippage: true,
+    isMarketOrderWidget: true,
+    recipient,
+    showRecipient,
+    isTradePriceUpdating: isRateLoading,
+    priceImpact,
+    disableQuotePolling: isConfirmOpen,
+  }
```

Comment:
slots, params, etc create a new object in each render, would this generate additional re-renders?

Reply:
No, it's actually works well, because:
1. The objects are not used in useEffect/useCallback/useMemo deps and they are destructured inside
2. The latest version of React digests them well

Just tried to add `useSafeMemoObject()` for the objects, no changes